mrmustard/utils/parameter.py

```diff
@@ -0,0 +1,164 @@
+# Copyright 2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+The classes in this module represent parameters passed to the
+quantum operations represented by :class:`~.Parametrized` subclasses.
+
+Parameter types
+---------------
+There are three basic types of parameters:
+1. **Numerical parameters** (bound and fixed): An immediate, immutable numerical python objects
+   (float, complex, int, numerical array). Implemented as-is, not encapsulated in a class.
+   They are dynamically assigned to properties of the class.
+2. **Trainable parameters** (bound but not fixed): These are parameters that are updated by
+   the optimization procedure. There are three types of trainable parameters which define the
+   optimization procedure: symplectic, euclidian and orthogonal.
+3. **Constant parameters** (bound and fixed): This class of parameters belong to the autograd
+   backend but remain fixed during the optimization procedure.
+"""
+# pylint: disable=super-init-not-called
+
+from abc import ABC, abstractmethod, abstractproperty
+
+from mrmustard.math import Math
+from mrmustard.types import Tensor, Tuple
+
+math = Math()
+
+
+class Parameter(ABC):
+    @abstractmethod
+    def __init__(self, value, name, owner=None) -> None:
+        pass
+
+    @property
+    def value(self) -> Tensor:
+        return self._value
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def owner(self) -> str:
+        return self._owner
+
+    @property
+    def type(self) -> str:
+        return self.__class__.__name__.lower()
+
+
+class Trainable(Parameter, ABC):
+    @abstractmethod
+    def __init__(self, value, bounds, name, owner=None) -> None:
+        pass
+
+    @abstractmethod
+    def update(self, grad, learning_rate) -> Tuple[Tensor, Tensor]:
+        pass
+
+
+class Symplectic(Trainable):
+    def __init__(self, value, bounds, name, owner=None) -> None:
```

Comment:
Symplectic and Orthogonal technically shouldn't support `bounds`.
Can we omit it from `__init__` and use `(None,None)` when calling `math.new_variable`?